### PR TITLE
Refactored Keccak hashes

### PR DESCRIPTION
**Pull request by batconjurer**
This PR
 * Moves the ABI encoding trait to a more appropriate location
 * Refactors the ABI encoding trait 
 * Adds useful traits to the Keccak hash type

### CODE REVIEW

File: shared/src/types/keccak.rs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+pub enum Error {
+    #[error("TEMPORARY error: {error}")]
+    Temporary { error: String },
+    #[error("Failed trying to convert slice to a hash: {0}")]
+    ConversionFailed(std::array::TryFromSliceError),
+    #[error("Failed to convert string into a hash: {0}")]
+    FromStringError(hex::FromHexError),
+}
+
+/// A Keccak hash
```
Comment (sug0) [inline shared/src/types/keccak.rs:24]:
```suggestion
/// Represents a Keccak hash.
```
not a fan of docstrings starting with articles

File: shared/src/types/keccak.rs [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+use std::convert::TryFrom;
+use std::fmt::Display;
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use hex::FromHex;
+use thiserror::Error;
+
+use crate::types::hash::{Hash, HASH_LENGTH};
+
+#[allow(missing_docs)]
```
Comment (sug0) [inline shared/src/types/keccak.rs:13]: add a docstring?

File: shared/src/types/keccak.rs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+impl Display for KeccakHash {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for byte in &self.0 {
+            write!(f, "{:02X}", byte)?;
+        }
+        Ok(())
+    }
+}
+
```
Comment (sug0) [inline shared/src/types/keccak.rs:46]:
```suggestion
```

**Comment by sug0** [comment]
CI is also complaining about `make fmt`

**Review by sug0 — APPROVED** [review]
ci still complains about docstrings. xd

File: shared/src/types/keccak.rs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        KeccakHash(hash.0)
+    }
+}
+
+impl TryFrom<&[u8]> for KeccakHash {
+    type Error = Error;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if value.len() != HASH_LENGTH {
+            return Err(Error::Temporary {
```
Comment (james-chf) [inline shared/src/types/keccak.rs:63]: This `Error::Temporary` should be replaced (e.g. with `Error::InvalidLength` or similar?). Or we could remove it and the `try_from` call below should catch this error case, I think

File: shared/src/types/keccak.rs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    }
+}
+
+impl TryFrom<String> for KeccakHash {
+    type Error = Error;
+
+    fn try_from(string: String) -> Result<Self, Error> {
+        let bytes: Vec<u8> =
+            Vec::from_hex(string).map_err(Error::FromStringError)?;
+        Self::try_from(bytes.as_slice())
```
Comment (james-chf) [inline shared/src/types/keccak.rs:83]: We could rely on the `impl TryFrom<&str>`
```suggestion
        string.as_str().try_into()
```

File: shared/src/types/keccak.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        if value.len() != HASH_LENGTH {
+            return Err(Error::Temporary {
+                error: format!(
+                    "Unexpected tx hash length {}, expected {}",
+                    value.len(),
+                    HASH_LENGTH
+                ),
+            });
+        }
+        let hash: [u8; 32] =
```
Comment (james-chf) [inline shared/src/types/keccak.rs:71]:
```suggestion
        let hash: [u8; HASH_LENGTH] =
```

File: shared/src/types/keccak.rs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use hex::FromHex;
+use thiserror::Error;
+
+use crate::types::hash::{Hash, HASH_LENGTH};
+
+/// Errors for converting / parsing Keccak hashes
+#[derive(Error, Debug)]
+pub enum Error {
```
Comment (james-chf) [inline shared/src/types/keccak.rs:15]: We could call this `TryFromError`?

**Review by james-chf — APPROVED** [review]

**Pull request merged**